### PR TITLE
Use typed_kwargs for dlang options

### PR DESCRIPTION
**Pull request by dcbaker**
I've also taken the liberty of doing a little cleanup and better type management internally by using a `TypedDict`. The big LoC change here is because I noticed that there's an exactly duplicate method implementation between and parent and child class, and I simply removed the child implementation

### CODE REVIEW

**Review by tristan957 — APPROVED** [review]

**Review by eli-schwartz — COMMENTED** [review]
>  compilers/d: use DFeatures for get_features_args 

commit message mentions removing the duplicate, though it was actually refactored into its own commit.

File: mesonbuild/interpreter/type_checking.py [hunk 1]
```diff
@@ -536,6 +536,7 @@ def link_whole_validator(values: T.List[T.Union[StaticLibrary, CustomTarget, Cus
     *_ALL_TARGET_KWS,
     BT_SOURCES_KW,
     RUST_CRATE_TYPE_KW,
+    KwargInfo('d_debug', ContainerTypeInfo(list, (str, int)), default=[], listify=True),
```
[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/interpreter/type_checking.py:540]: hm?

I just realized that we have the python `bool is int` problem leaking through here...

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply mesonbuild/interpreter/type_checking.py:540]: Yeah. It's horrible.

File: mesonbuild/interpreter/type_checking.py [hunk 2]
```diff
@@ -539,6 +539,7 @@ def link_whole_validator(values: T.List[T.Union[StaticLibrary, CustomTarget, Cus
     RUST_CRATE_TYPE_KW,
     KwargInfo('d_debug', ContainerTypeInfo(list, (str, int)), default=[], listify=True),
     D_MODULE_VERSIONS_KW,
+    KwargInfo('d_unittest', bool, default=False),
```
[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/interpreter/type_checking.py:542]: Nope, right now python's "truthiness" is leaking through

**Pull request merged**